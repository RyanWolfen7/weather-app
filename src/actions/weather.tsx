import { FETCH_WEATHER, FETCH_WEATHER_SUCCESS, FETCH_WEATHER_FAILURE, CLEAR_WEATHER } from '../types';

const fetchWeather = (city: string) => ({
  types: [FETCH_WEATHER, FETCH_WEATHER_SUCCESS, FETCH_WEATHER_FAILURE],
  payload: {
    request: {
      url: `/forecast/daily?city=${city}`,
      method: 'GET',
      headers: {}
    }
  }
});

const clearWeather = () => ({ type: CLEAR_WEATHER });

export default {
  fetchWeather,
  clearWeather,
};
