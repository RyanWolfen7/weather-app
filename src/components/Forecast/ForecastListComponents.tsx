import styled from 'styled-components';

export const ErrorMessage = styled.span`
  font-size: 30px;
`;

export const ErrorWrapper = styled.div`
  justify-content: center;
  align-items: center;
  display: flex;
  flex-direction: column;
  height: 100vh;
`;

export const TextTheme = styled.span`
  color: #333;
  font-weight: 100;
  text-align: center;
`;

export const Title = styled(TextTheme)`
  font-size: 3rem;
  margin-top: 30px;
`;

export const Subtitle = styled(TextTheme)`
  font-size: 1rem;
`;

export const ForecastDayImage = styled.img`
  height: 130px;
`;

export const ForecastDay = styled.div`
  width: 30%;
  display: flex;
  margin: 10px;
  flex-direction: column;
`;

export const ForecastDays = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const ForecastSubheader = styled(TextTheme)`
  font-size: 2em;
`;

export const ForecastCaption = styled(TextTheme)`
  font-size: 1em;
`;

export const Wrapper = styled.div`
  align-items: center;
  justify-content: space-around;
  display: flex;
  flex-direction: column;
  flex: 1;
  height: 100%;
`;

export const FreeArea = styled.div`
  height: ${(props: { height: number }) => props.height}px;
`;
