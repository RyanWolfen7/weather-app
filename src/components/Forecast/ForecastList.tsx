import React from 'react';
import { useSelector } from 'react-redux';
import { withRouter, RouteComponentProps } from 'react-router';
import { sentenceCase } from 'sentence-case';
import { ErrorMessage, ErrorWrapper, Title, Subtitle, ForecastDayImage, ForecastDay, ForecastDays, Wrapper, FreeArea, ForecastSubheader, ForecastCaption } from './ForecastListComponents';
import Search from '../Search/Search';
import Weather from '../../interfaces/Weather';
import { dateAt } from '../../helpers';

interface ForecastListInterface extends RouteComponentProps<{ city: string }> { }

const ForecastList: React.FC<ForecastListInterface> = props => {

  const { data, error } = useSelector((state: { weather: Weather }) => state.weather);

  const goToDetails = (id: number) => {
    props.history.push(`/details/${props.match.params.city}/${id}`);
  }

  return (
    <>
      {error && <ErrorWrapper>
        <ErrorMessage>{sentenceCase(error.message)}</ErrorMessage>
        <Search />
      </ErrorWrapper>}
      {
        !!data.city && <Wrapper>
          <Title>{sentenceCase(data.city.name)}, {data.city.country}</Title>
          <Subtitle>Population: {data.city.population}</Subtitle>
          <FreeArea height={50} />
          <ForecastDays>
            {(data.list || []).map((el, index) => <ForecastDay onClick={() => goToDetails(el.weather[0].id)} key={index}>
              <ForecastDayImage src={`/assets/weather-icons/${el.weather[0].icon}.svg`} />
              <ForecastSubheader>{dateAt(index)}</ForecastSubheader>
              <ForecastCaption>{el.weather[0].main}</ForecastCaption>
            </ForecastDay>)}
          </ForecastDays>
        </Wrapper>
      }
    </>
  );
}

export default withRouter(ForecastList);
