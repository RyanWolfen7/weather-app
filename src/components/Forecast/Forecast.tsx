import React, { useEffect, useState } from 'react';
import { withRouter, RouteComponentProps } from 'react-router';
import { useDispatch, useSelector } from 'react-redux';
import { Fade, Container, Spinner } from 'reactstrap';
import styled from 'styled-components';
import actions from '../../actions';
import ForecastList from './ForecastList';
import Weather from '../../interfaces/Weather';

interface ForecastInterface extends RouteComponentProps<{ city: string }> { }

const Forecast: React.FC<ForecastInterface> = ({ match }) => {
  const [fadeIn, setFadeIn] = useState(false);
  const { isLoading, data, error } = useSelector((state: { weather: Weather }) => state.weather);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!data.city && !error) dispatch(actions.weather.fetchWeather(match.params.city));
  }, [error, data, match.params, dispatch]);

  useEffect(() => {
    setFadeIn(true);
  }, [data.city]);

  return (
    <Parent>
      <Container>
        {isLoading && <Spinner width={3} height={3} type="grow" color="warning" />}
        {!isLoading && <FadeWrapper in={fadeIn} tag="div">
          <ForecastList />
        </FadeWrapper>}
      </Container>
    </Parent>
  );
}

export default withRouter(Forecast);

const Parent = styled.div`
  background-color: white;
`;

const FadeWrapper = styled(Fade)`
  height: 100%;
`;
