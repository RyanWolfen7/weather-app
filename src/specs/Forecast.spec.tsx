import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import Forecast from '../components/Forecast/Forecast';
import WeatherInterface, { DataResponse } from '../interfaces/Weather';
import dataMock from './mocks/data';
import actions from '../actions/weather';
import 'jest-styled-components';

let wrapper: ReactWrapper;
const useDispatch = jest.fn();
let useData = false;

const useSelectorWithData: (allowData: boolean) => { weather: WeatherInterface } = allowData => ({
  weather: {
    isLoading: false,
    data: allowData ? dataMock : {} as DataResponse,
    error: null
  }
});

jest.mock('react-redux', () => ({
  useDispatch: () => useDispatch,
  useSelector: (cb: Function) => cb(useSelectorWithData(useData))
}));

beforeEach(() => {
  wrapper = mount(
    <MemoryRouter initialEntries={['/forecast/cambridge']} keyLength={0}>
      <Route path='/forecast/:city'>
        <Forecast />
      </Route>
    </MemoryRouter>
  );
});

test('Get Forecast by city name', () => {
  expect(useDispatch).toBeCalledWith(actions.fetchWeather('cambridge'));
  useData = true;
});

test('Display Forecast by city name', () => {
  expect(wrapper.text()).toMatch('Cambridge, CA');
});

test('Forecast to match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});
