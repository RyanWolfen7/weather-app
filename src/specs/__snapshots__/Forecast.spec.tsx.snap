// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast to match snapshot 1`] = `
.c3 {
  color: #333;
  font-weight: 100;
  text-align: center;
  font-size: 3rem;
  margin-top: 30px;
}

.c4 {
  color: #333;
  font-weight: 100;
  text-align: center;
  font-size: 1rem;
}

.c8 {
  height: 130px;
}

.c7 {
  width: 30%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c9 {
  color: #333;
  font-weight: 100;
  text-align: center;
  font-size: 2em;
}

.c10 {
  color: #333;
  font-weight: 100;
  text-align: center;
  font-size: 1em;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
}

.c5 {
  height: 50px;
}

.c0 {
  background-color: white;
}

.c1 {
  height: 100%;
}

<MemoryRouter
  initialEntries={
    Array [
      "/forecast/cambridge",
    ]
  }
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/forecast/cambridge",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/forecast/cambridge",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      path="/forecast/:city"
    >
      <withRouter(Forecast)>
        <Forecast
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "pathname": "/forecast/cambridge",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/forecast/cambridge",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/forecast/cambridge",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {
                "city": "cambridge",
              },
              "path": "/forecast/:city",
              "url": "/forecast/cambridge",
            }
          }
        >
          <styled.div>
            <div
              className="c0"
            >
              <Container
                tag="div"
              >
                <div
                  className="container"
                >
                  <Styled(Fade)
                    in={true}
                    tag="div"
                  >
                    <Fade
                      appear={true}
                      baseClass="fade"
                      baseClassActive="show"
                      className="c1"
                      enter={true}
                      exit={true}
                      in={true}
                      mountOnEnter={false}
                      onEnter={[Function]}
                      onEntered={[Function]}
                      onEntering={[Function]}
                      onExit={[Function]}
                      onExited={[Function]}
                      onExiting={[Function]}
                      tag="div"
                      timeout={150}
                      unmountOnExit={false}
                    >
                      <Transition
                        appear={true}
                        enter={true}
                        exit={true}
                        in={true}
                        mountOnEnter={false}
                        onEnter={[Function]}
                        onEntered={[Function]}
                        onEntering={[Function]}
                        onExit={[Function]}
                        onExited={[Function]}
                        onExiting={[Function]}
                        timeout={150}
                        unmountOnExit={false}
                      >
                        <div
                          className="c1 fade"
                        >
                          <withRouter(ForecastList)>
                            <ForecastList
                              history={
                                Object {
                                  "action": "POP",
                                  "block": [Function],
                                  "canGo": [Function],
                                  "createHref": [Function],
                                  "entries": Array [
                                    Object {
                                      "hash": "",
                                      "pathname": "/forecast/cambridge",
                                      "search": "",
                                      "state": undefined,
                                    },
                                  ],
                                  "go": [Function],
                                  "goBack": [Function],
                                  "goForward": [Function],
                                  "index": 0,
                                  "length": 1,
                                  "listen": [Function],
                                  "location": Object {
                                    "hash": "",
                                    "pathname": "/forecast/cambridge",
                                    "search": "",
                                    "state": undefined,
                                  },
                                  "push": [Function],
                                  "replace": [Function],
                                }
                              }
                              location={
                                Object {
                                  "hash": "",
                                  "pathname": "/forecast/cambridge",
                                  "search": "",
                                  "state": undefined,
                                }
                              }
                              match={
                                Object {
                                  "isExact": true,
                                  "params": Object {
                                    "city": "cambridge",
                                  },
                                  "path": "/forecast/:city",
                                  "url": "/forecast/cambridge",
                                }
                              }
                            >
                              <styled.div>
                                <div
                                  className="c2"
                                >
                                  <Styled(styled.span)>
                                    <span
                                      className="sc-AykKE c3"
                                    >
                                      Cambridge
                                      , 
                                      CA
                                    </span>
                                  </Styled(styled.span)>
                                  <Styled(styled.span)>
                                    <span
                                      className="sc-AykKE c4"
                                    >
                                      Population: 
                                      120372
                                    </span>
                                  </Styled(styled.span)>
                                  <styled.div
                                    height={50}
                                  >
                                    <div
                                      className="c5"
                                      height={50}
                                    />
                                  </styled.div>
                                  <styled.div>
                                    <div
                                      className="c6"
                                    >
                                      <styled.div
                                        key="0"
                                        onClick={[Function]}
                                      >
                                        <div
                                          className="c7"
                                          onClick={[Function]}
                                        >
                                          <styled.img
                                            src="/assets/weather-icons/10d.svg"
                                          >
                                            <img
                                              className="c8"
                                              src="/assets/weather-icons/10d.svg"
                                            />
                                          </styled.img>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c9"
                                            >
                                              Sun Feb 02
                                            </span>
                                          </Styled(styled.span)>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c10"
                                            >
                                              Rain
                                            </span>
                                          </Styled(styled.span)>
                                        </div>
                                      </styled.div>
                                      <styled.div
                                        key="1"
                                        onClick={[Function]}
                                      >
                                        <div
                                          className="c7"
                                          onClick={[Function]}
                                        >
                                          <styled.img
                                            src="/assets/weather-icons/10d.svg"
                                          >
                                            <img
                                              className="c8"
                                              src="/assets/weather-icons/10d.svg"
                                            />
                                          </styled.img>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c9"
                                            >
                                              Mon Feb 03
                                            </span>
                                          </Styled(styled.span)>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c10"
                                            >
                                              Rain
                                            </span>
                                          </Styled(styled.span)>
                                        </div>
                                      </styled.div>
                                      <styled.div
                                        key="2"
                                        onClick={[Function]}
                                      >
                                        <div
                                          className="c7"
                                          onClick={[Function]}
                                        >
                                          <styled.img
                                            src="/assets/weather-icons/13d.svg"
                                          >
                                            <img
                                              className="c8"
                                              src="/assets/weather-icons/13d.svg"
                                            />
                                          </styled.img>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c9"
                                            >
                                              Tue Feb 04
                                            </span>
                                          </Styled(styled.span)>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c10"
                                            >
                                              Snow
                                            </span>
                                          </Styled(styled.span)>
                                        </div>
                                      </styled.div>
                                      <styled.div
                                        key="3"
                                        onClick={[Function]}
                                      >
                                        <div
                                          className="c7"
                                          onClick={[Function]}
                                        >
                                          <styled.img
                                            src="/assets/weather-icons/04d.svg"
                                          >
                                            <img
                                              className="c8"
                                              src="/assets/weather-icons/04d.svg"
                                            />
                                          </styled.img>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c9"
                                            >
                                              Wed Feb 05
                                            </span>
                                          </Styled(styled.span)>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c10"
                                            >
                                              Clouds
                                            </span>
                                          </Styled(styled.span)>
                                        </div>
                                      </styled.div>
                                      <styled.div
                                        key="4"
                                        onClick={[Function]}
                                      >
                                        <div
                                          className="c7"
                                          onClick={[Function]}
                                        >
                                          <styled.img
                                            src="/assets/weather-icons/03d.svg"
                                          >
                                            <img
                                              className="c8"
                                              src="/assets/weather-icons/03d.svg"
                                            />
                                          </styled.img>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c9"
                                            >
                                              Thu Feb 06
                                            </span>
                                          </Styled(styled.span)>
                                          <Styled(styled.span)>
                                            <span
                                              className="sc-AykKE c10"
                                            >
                                              Clouds
                                            </span>
                                          </Styled(styled.span)>
                                        </div>
                                      </styled.div>
                                    </div>
                                  </styled.div>
                                </div>
                              </styled.div>
                            </ForecastList>
                          </withRouter(ForecastList)>
                        </div>
                      </Transition>
                    </Fade>
                  </Styled(Fade)>
                </div>
              </Container>
            </div>
          </styled.div>
        </Forecast>
      </withRouter(Forecast)>
    </Route>
  </Router>
</MemoryRouter>
`;
